src/relay/libp2p_relay_server.erl

```diff
@@ -105,73 +115,73 @@ handle_cast(init_relay, #state{started=false, swarm=Swarm}=State0) ->
                     lager:debug("joined peerbook ~p notifications", [Peerbook]),
                     Peers0 = libp2p_peerbook:values(Peerbook),
                     lists:filter(fun(E) ->
-                                         libp2p_peer:pubkey_bin(E) /= SwarmAddr
-                                 end, Peers0);
+                        libp2p_peer:pubkey_bin(E) /= SwarmAddr
+                    end, Peers0);
                 _ ->
                     State0#state.peers
             end,
-    State = State0#state{peers=sort_peers(Peers, SwarmAddr), address=SwarmAddr},
-    case int_relay(State) of
-        {ok, _} ->
-            lager:debug("relay started successfuly"),
-            {noreply, add_flap(State#state{started=true})};
+    State = State0#state{peers=sort_peers(Peers, SwarmAddr)},
+    case init_relay(State) of
+        {ok, Pid} ->
+            _ = erlang:monitor(process, Pid),
+            lager:info("relay started successfuly with ~p", [Pid]),
+            {noreply, add_flap(State#state{stream=Pid, address=undefined, retrying=undefined})};
         _Error ->
             lager:warning("could not initiate relay ~p", [_Error]),
-            erlang:send_after(2500, self(), try_relay),
-            {noreply, next_peer(State)}
+            {noreply, next_peer(retry(State))}
     end;
-handle_cast(init_relay, State) ->
+handle_cast(init_relay,  #state{stream=Pid}=State) when is_pid(Pid) ->
+    lager:info("requested to init relay but we already have one @ ~p", [Pid]),
     {noreply, State};
-handle_cast(connection_lost, State) ->
-    lager:debug("relay connection lost"),
-    self() ! try_relay,
-    {noreply, State#state{started=false, connection=undefined}};
 handle_cast(_Msg, State) ->
-    lager:debug("rcvd unknown cast msg: ~p", [_Msg]),
+    lager:warning("rcvd unknown cast msg: ~p", [_Msg]),
     {noreply, State}.
 
-handle_info({new_peers, NewPeers}, #state{started=true}=State) ->
-    {noreply, State#state{peers=sort_peers(merge_peers(NewPeers, State#state.peers), State#state.address), peer_index=1}};
-handle_info({new_peers, NewPeers}, #state{started=false}=State) ->
-    self() ! try_relay,
-    {noreply, State#state{peers=sort_peers(merge_peers(NewPeers, State#state.peers), State#state.address)}};
-handle_info(try_relay, State = #state{started=false}) ->
-    case int_relay(State) of
-        {ok, Connection} ->
-            lager:debug("relay started successfuly"),
-            {noreply, add_flap(State#state{started=true, connection=Connection})};
-        {error, no_peer} ->
-            lager:warning("could not initiate relay no peer found"),
-            {noreply, State};
-        {error, no_address} ->
-            lager:warning("could not initiate relay, swarm has no listen address"),
-            {noreply, State};
+handle_info({new_peers, NewPeers}, #state{tid=TID, stream=Pid}=State) when is_pid(Pid) ->
+    Swarm = libp2p_swarm:swarm(TID),
+    SwarmAddr = libp2p_swarm:pubkey_bin(Swarm),
```

Comment:
Technically it's not the SwarmAddr, but the SwarmPubKey? No big deal but the name got me confused with it's p2p address for a sec